election-server/src/bin/indexer.rs

```diff
@@ -0,0 +1,605 @@
+/// any transactions made to the contract due to either listening to
+/// transactions made to the wrong contract of a wrong contract entrypoint.
+async fn verify_contract(
+    node: &mut Client,
+    contract_address: &ContractAddress,
+) -> anyhow::Result<()> {
+    let instance_info = node
+        .get_instance_info(*contract_address, BlockIdentifier::LastFinal)
+        .await
+        .context("Could not get instance info for election contract")?
+        .response;
+    let methods = match instance_info {
+        InstanceInfo::V0 { .. } => return Err(anyhow!("Expected V1 contract")),
+        InstanceInfo::V1 { methods, .. } => methods,
+    };
+
+    anyhow::ensure!(
+        methods.iter().any(|m| m == REGISTER_VOTES_RECEIVE),
+        "Expected method with receive name \"{}\" to be available on contract",
+        REGISTER_VOTES_RECEIVE
+    );
+
+    Ok(())
+}
+
+/// Find the block height corresponding to the start time of the election. If
+/// the election start time is in the future, this function will pause the
+/// thread until the election has started, after which it will return the block
+/// height corresponding to the latest finalized block.
+async fn find_election_start_height(
+    client: &mut Client,
+    contract_address: &ContractAddress,
+) -> anyhow::Result<AbsoluteBlockHeight> {
+    let context = ContractContext::new(
+        *contract_address,
+        OwnedReceiveName::new_unchecked(CONFIG_VIEW.to_string()),
+    );
+    let InvokeContractResult::Success {
+        return_value: Some(election_config),
+        ..
+    } = client
+        .invoke_instance(BlockIdentifier::LastFinal, &context)
+        .await?
+        .response
+    else {
+        return Err(anyhow!("Expected to be able to query election config"));
+    };
+    let election_config: ElectionConfig =
+        contracts_common::from_bytes(election_config.value.as_ref())
+            .context("Failed to parse election config from contract invocation result")?;
+    let election_start: DateTime<Utc> = election_config.election_start.try_into()?;
+
+    let now = Utc::now();
+    if election_start > now {
+        tracing::info!(
+            "Election has not started yet. Resuming execution at {}",
+            election_start
+        );
+
+        // As there is nothing to do until the election starts, wait until then.
+        sleep(election_start.signed_duration_since(now).to_std()?).await;
+    } else {
+    }
+    let (creation_height, ..) = client
+        .find_instance_creation(.., *contract_address)
+        .await
+        .context("Could not find contract instance creation block")?;
+
+    let result = client
+        .find_first_finalized_block_no_later_than(creation_height.., election_start)
+        .await;
+
+    let block_info = if let Ok(block_info) = result {
+        block_info
+    } else {
+        // Fall back to height of last finalized block
+        client
+            .get_block_info(BlockIdentifier::LastFinal)
```

Comment:
This can in principle be wrong due to time delays.
Since you did `InvokeInstance` at the beginning with the last finalized block, we should never reach this branch.

If we do, I think we should just fail.

Reply:
Yes this was a bit annoying I remember. It is probably best to just fail.. Not including a ballot due to this would not be the best 😛